src/utxo/db.rs

```diff
@@ -0,0 +1,132 @@
+        let mut options = Options::default();
+        options.set_compression_type(DBCompressionType::Snappy);
+        options.create_if_missing(true);
+        let db = DB::open(&options, path)?;
+
+        let updated_up_to_height = db
+            .get("updated_up_to_height")?
+            .map(|e| e.try_into().unwrap())
+            .map(|e| i32::from_ne_bytes(e))
+            .unwrap_or(-1);
+
+        Ok(DbUtxo {
+            db,
+            updated_up_to_height,
+            inserted_outputs: 0,
+        })
+    }
+}
+
+impl Utxo for DbUtxo {
+    fn add(&mut self, block: &Block, height: u32) {
+        let height = height as i32;
+        debug!(
+            "height: {} updated_up_to: {}",
+            height, self.updated_up_to_height
+        );
+        if height > self.updated_up_to_height {
+            // since we can spend outputs created in this same block, we first put outputs in memory...
+            let total_outputs = block.txdata.iter().map(|e| e.output.len()).sum();
+            let mut current_outputs = HashMap::with_capacity(total_outputs);
+            for tx in block.txdata.iter() {
+                let txid = tx.txid();
+                for (i, output) in tx.output.iter().enumerate() {
+                    let outpoint = OutPoint::new(txid, i as u32);
+                    current_outputs.insert(outpoint, output);
+                }
+            }
+
+            let total_inputs = block.txdata.iter().skip(1).map(|e| e.input.len()).sum();
+            let mut prevouts = Vec::with_capacity(total_inputs);
+            let mut batch = WriteBatch::default();
+            for tx in block.txdata.iter().skip(1) {
+                for input in tx.input.iter() {
+                    //...then we first check if inputs spend output created in this block
+                    match current_outputs.remove(&input.previous_output) {
+                        Some(tx_out) => {
+                            // we avoid touching the db entirely if it's spent in the same block
+                            prevouts.push(tx_out.clone())
+                        }
+                        None => {
+                            let key = input.previous_output.to_key();
+                            let tx_out = deserialize(&self.db.get(&key).unwrap().unwrap()).unwrap();
+                            batch.delete(&key);
+                            prevouts.push(tx_out);
+                            self.inserted_outputs += 1;
+                        }
+                    }
+                }
+            }
+
+            // and we put all the remaining outputs in db
+            for (k, v) in current_outputs.drain() {
+                batch.put(&k.to_key(), serialize(v));
+            }
+            batch.put(height.to_ne_bytes(), serialize(&prevouts));
+            batch.put("updated_up_to_height", height.to_ne_bytes());
+            self.db.write(batch).unwrap(); // TODO unwrap
```

Comment:
This write could be easily ofloaded to another thread by sending it via channel, and then `prevout` returned to caller. This way the caller can be already doing it's logic while the DB is writting stuff out. You might need to add some `fn flush()` on a `trait Utxo` to make sure the data have hit the storage and any errors bubbled out.

Reply:
Well, since it's used in a pipeline of threads already and that thread doesn't seem to be doing much else, maybe it's not useful.